pkg/parser/translate.go

```diff
@@ -419,21 +418,52 @@ func fromUDPIngressV1Alpha1(log logrus.FieldLogger, ingressList []*v1alpha1.UDPI
 			"udpingress_name":      ingress.Name,
 		})
 
-		result.ServiceNameToServices[ingress.Namespace+"."+ingress.Name] = kongstate.Service{
-			Service: kong.Service{
-				Name:     kong.String(ingress.Namespace + "." + ingress.Name),
-				Protocol: kong.String("udp"),
-				Host:     kong.String(ingressSpec.Host),
-				Port:     kong.Int(ingress.Spec.TargetPort),
-			},
-			Routes: []kongstate.Route{
-				{
-					Route: kong.Route{
-						Protocols:    []*string{kong.String("udp")},
-						Destinations: []*kong.CIDRPort{{Port: kong.Int(ingressSpec.ListenPort)}},
-					},
+		for i, rule := range ingressSpec.Rules {
+			// validate the ports and servicenames for the rule
+			if !util.IsValidPort(rule.Port) {
+				log.Errorf("invalid TCPIngress: invalid port: %d", rule.Port)
+				continue
+			}
+			if rule.Backend.ServiceName == "" {
+				log.Errorf("invalid UDPIngress: empty serviceName")
+				continue
+			}
+			if !util.IsValidPort(rule.Backend.ServicePort) {
+				log.Errorf("invalid UDPIngress: invalid servicePort: %d", rule.Backend.ServicePort)
+				continue
+			}
+
+			// generate the kong Route based on the listen port
+			route := kongstate.Route{
+				Ingress: util.FromK8sObject(ingress),
+				Route: kong.Route{
+					Name:         kong.String("udpingress" + "." + ingress.Namespace + "." + ingress.Name + "." + strconv.Itoa(i)),
+					Protocols:    kong.StringSlice("udp"),
+					Destinations: []*kong.CIDRPort{{Port: kong.Int(rule.Port)}},
 				},
-			},
+			}
+
+			// generate the kong Service backend for the UDPIngress rules
+			host := fmt.Sprintf("%s.%s.%d.svc", rule.Backend.ServiceName, ingress.Namespace, rule.Backend.ServicePort)
+			serviceName := fmt.Sprintf("%s.%s.%d", ingress.Namespace, rule.Backend.ServiceName, rule.Backend.ServicePort)
```

Comment:
This can create a conflict. We need to include the protocol for UDP ports.

A TCPIngress and UDPIngress that point to the same port on the same Service not only result in one Kong service, but also make the associated TCPIngress route disappear. For demonstration, see [this set of example resources](https://gist.github.com/rainest/aff37cc4216775d29328ceab4a62917c). Ignore that the Deployment doesn't actually support the additional ports, we just need it to fill in endpoints.

TCPIngress/Ingress aren't affected AFAIK, since they can't share port integers--you just have to split those up even if you specify an appProtocol.

Depending on how complicated it is to fix this, we may be able to defer this for now. It's something of an edge case where we could reasonably say "don't do this, configure your way around it" initially.

Reply:
+1 to Travis's comment, and a request to make this one blocking because (IIUC) this shouldn't be that difficult to fix - just ensure that you generate object names from a space that doesn't overlap with the spaces for objects already generated elsewhere.

I think our lives will be easier (and weekends longer) if we avoid introducing naming conflicts like this when we have an opportunity to.

Reply:
:+1: 1ebf249f324d72504391101f23285e41bac9d831 is intended to correct for this